db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectivityUtil.java

```diff
@@ -130,14 +132,73 @@ public static double estimateJoinSelectivity(Optimizable innerTable, Conglomerat
             }
         }
         double selectivity = 1.d;
+        List<JoinPredicateSelectivity> predSelectivities = new ArrayList<>();
         if (predList != null) {
             for (int i = 0; i < predList.size(); i++) {
                 Predicate p = (Predicate) predList.getOptPredicate(i);
                 if (!isTheRightJoinPredicate(p, predicateType))
                     continue;
 
-                selectivity *= p.joinSelectivity(innerTable, innerCD, innerRowCount, outerRowCount, selectivityJoinType);
+                predSelectivities.add(new JoinPredicateSelectivity(p, innerTable, QualifierPhase.JOIN,
+                                                           p.joinSelectivity(innerTable, innerCD, innerRowCount,
+                                                                             outerRowCount, selectivityJoinType)));
+            }
+        }
+        if (predSelectivities.size() == 1)
+            selectivity = predSelectivities.get(0).getSelectivity();
+        else {
+            // Sort the selectivities and combine them using computeSqrtLevel
+            // as is done in ScanCostFunction.
+            Collections.sort(predSelectivities);
+
+            // Map from predicate number (in the order added to the map) to the
+            // selectivity of the predicate.
+            Map<Integer, JoinPredicateSelectivity> selectivityMap = new TreeMap<>();
+
+            // Map from column number to index into selectivityMap.
+            // This double lookup is done so that the main data structure
+            // holding selectivities has no duplicates.
+            Map<Integer,Integer> selectivityIndexMap = new HashMap<>();
+
+            int index = 0;
+            for (JoinPredicateSelectivity predicateSelectivity:predSelectivities) {
+                Predicate p = predicateSelectivity.getPredicate();
+                int tableNumber = innerTable.getTableNumber();
+                Set<Integer> columnSet = p.getReferencedColumns().get(tableNumber);
+                ReferencedColumnsMap referencedColumnsMap = p.getReferencedColumns();
+
+                Integer mapIndex = null;
+                if (columnSet != null)
+                    for (Integer I:columnSet) {
+                        mapIndex = selectivityIndexMap.get(I);
+                        if (mapIndex != null)
+                            break;
+                    }
```

Comment:
Maybe a pair of curly braces for the if block, since it contains multiple lines.

Reply:
Added braces.